compiler/IREmitter/NameBasedIntrinsics.cc

```diff
@@ -82,29 +82,34 @@ class DefineClassIntrinsic : public NameBasedIntrinsicMethod {
             return Payload::rubyNil(cs, builder);
         }
 
-        // this is wrong and will not work for `class <<self`
-        auto classNameCStr = Payload::toCString(cs, IREmitterHelpers::showClassNameWithoutOwner(cs, sym), builder);
-        auto isModule = sym.data(cs)->superClass() == core::Symbols::Module();
-        auto funcSym = cs.gs.lookupStaticInitForClass(attachedClass);
-
         llvm::Value *module = nullptr;
 
-        if (!IREmitterHelpers::isRootishSymbol(cs, sym.data(cs)->owner)) {
-            auto getOwner = Payload::getRubyConstant(cs, sym.data(cs)->owner, builder);
-            if (isModule) {
-                module = builder.CreateCall(cs.getFunction("sorbet_defineNestedModule"), {getOwner, classNameCStr});
+        auto funcSym = cs.gs.lookupStaticInitForClass(attachedClass);
+        if (!attachedClass.data(cs)->isSingletonClass(cs)) {
+            auto classNameCStr = Payload::toCString(cs, IREmitterHelpers::showClassNameWithoutOwner(cs, sym), builder);
+            auto isModule = sym.data(cs)->superClass() == core::Symbols::Module();
+
+            if (!IREmitterHelpers::isRootishSymbol(cs, sym.data(cs)->owner)) {
+                auto getOwner = Payload::getRubyConstant(cs, sym.data(cs)->owner, builder);
+                if (isModule) {
+                    module = builder.CreateCall(cs.getFunction("sorbet_defineNestedModule"), {getOwner, classNameCStr});
+                } else {
+                    auto rawCall = Payload::getRubyConstant(cs, sym.data(cs)->superClass(), builder);
+                    module = builder.CreateCall(cs.getFunction("sorbet_defineNestedClass"),
+                                                {getOwner, classNameCStr, rawCall});
+                }
             } else {
-                auto rawCall = Payload::getRubyConstant(cs, sym.data(cs)->superClass(), builder);
-                module =
-                    builder.CreateCall(cs.getFunction("sorbet_defineNestedClass"), {getOwner, classNameCStr, rawCall});
+                if (isModule) {
+                    module = builder.CreateCall(cs.getFunction("sorbet_defineTopLevelModule"), {classNameCStr});
+                } else {
+                    auto rawCall = Payload::getRubyConstant(cs, sym.data(cs)->superClass(), builder);
+                    module =
+                        builder.CreateCall(cs.getFunction("sorbet_defineTopClassOrModule"), {classNameCStr, rawCall});
+                }
```

Comment:
Just voicing my understanding out loud in case it's wrong: all the green code in this block is pre-existing, it was just moved inside the `if(!attachedClass.data(cs)->isSingletonClass(cs))`.

Reply:
This is a correct summary of the situation!